### PR TITLE
Core/Player: implement ResetCoolDownAfterDuel configurable feature

### PR DESCRIPTION
**Pull request by FrancescoBorzi**
Reset the cooldown of both players after duel. Disabled by default.
- In order to prevent abuse, it does the reset only if the player didn't have any cooldown before starting the duel
- It considers the _arena cooldowns_ only (all cooldowns that last less than 10 minutes)

### CODE REVIEW

File: src/server/game/Entities/Player/Player.cpp [hunk 1]
```diff
@@ -7610,6 +7610,12 @@ void Player::DuelComplete(DuelCompleteType type)
     duel->opponent->SetGuidValue(PLAYER_DUEL_ARBITER, ObjectGuid::Empty);
     duel->opponent->SetUInt32Value(PLAYER_DUEL_TEAM, 0);
 
+    if (sWorld->getIntConfig(CONFIG_RESET_COOLDOWN_AFTER_DUEL) == 1)
+    {
+        GetSpellHistory()->ResetAllCooldowns();
+        duel->opponent->GetSpellHistory()->ResetAllCooldowns();
```
Comment (joschiwald) [inline src/server/game/Entities/Player/Player.cpp]: profession cooldowns? ^^

**Comment by ShinDarth** [comment]
Updated, tested and working.

**Comment by digz6666** [comment]
:+1:

File: src/server/game/Entities/Player/Player.h [hunk 2]
```diff
@@ -1946,6 +1947,10 @@ class Player : public Unit, public GridObject<Player>
         void SetHonorPoints(uint32 value);
         void SetArenaPoints(uint32 value);
 
+        bool m_hasCoolDownBeforeDuel;
```
Comment (Takenbacon) [inline src/server/game/Entities/Player/Player.h]: This needs to be initialized in player constructor.

**Comment by MitchesD** [comment]
is that blizzlike or custom?

File: src/server/game/Entities/Player/Player.cpp [hunk 3]
```diff
@@ -7610,6 +7610,21 @@ void Player::DuelComplete(DuelCompleteType type)
     duel->opponent->SetGuidValue(PLAYER_DUEL_ARBITER, ObjectGuid::Empty);
     duel->opponent->SetUInt32Value(PLAYER_DUEL_TEAM, 0);
 
+    if (sWorld->getBoolConfig(CONFIG_RESET_COOLDOWN_AFTER_DUEL) &&
+        type != DUEL_INTERRUPTED)
+    {
+        std::string hadCoolDownBeforeDuelText = "You had cooldowns before starting the duel, so your cooldowns haven't been reset.";
```
Comment (Shauren) [inline src/server/game/Entities/Player/Player.cpp]: Text needs to be trinity_string'ed for localization

**Comment by MitchesD** [comment]
I wonder if we can move these custom changes somewhere else, to don't make Player class even more bigger, it could make that class a bit smaller if we move custom things to somewhere else and just call them here

**Comment by joschiwald** [comment]
@MitchesD PlayerScripts

**Comment by ShinDarth** [comment]
@Shauren  @Takenbacon done

@MitchesD it's just few lines inside Player.cpp

File: src/server/game/Entities/Player/Player.h (hunk 2 above)
Comment (ShinDarth) [inline src/server/game/Entities/Player/Player.h]: done

**Comment by L30m4nc3r** [comment]
@ShinDarth make 2 version.
1) After Duel.
2) Before Duel.
-- You have cooldowns, /duel
-- Duel Timer 3, 2, 1, DUEL_START_COOLDOWN_RESET

For me work this code:
    void OnDuelStart(Player* player, Player* plTarget)// override
    {
        for (uint32 phase = 2; phase <= ULONG_MAX / 2; phase *= 2)
        {
        RemoveArenaSpellCooldowns(true);
        return;
        }
    }
But i have micro-lag after 1 second.

**Comment by ShinDarth** [comment]
@L30m4nc3r that would nullify my check to prevent abuse of this.

**Comment by Rochet2** [comment]
> But i have micro-lag after 1 second.

@L30m4nc3r The for-loop is causing the lag and looks to be unnecessary, unless that is not the whole code.

**Comment by Keader** [comment]
If I use Divine Shield before duel  and use stun(Hammer of Justice) during the duel, the right would not be the hammer of justice have reset cooldown?
The current form, with any cooldown before the duel, no spell has reset cooldown.

(sorry bad english)

**Comment by ShinDarth** [comment]
@Keader yes,  it is that way in order to prevent any abuse, as I wrote in the PR description

**Pull request merged**